<!-- Load Required Scripts -->
<script src="../dist/js/d3.min.js"></script>
<script src="../dist/js/d3.phylogram.js"></script>
<script src="../dist/js/newick.js"></script>
<script src="../dist/js/genomecomparevis.js"></script>
<link rel="stylesheet" href="../dist/css/linearplot.css" />


<!-- Content Header (Page header) -->
<section class="content-header">
    <h1>
        Results
    </h1>
    <ol class="breadcrumb">
        <li class="active"><a href="#"><i class="fa fa-bar-chart"></i> Results</a></li>
    </ol>
</section>

<!-- Main content -->
<section class="content">
    <div class="row">
        <div class="col-xs-12">
            <div class="box">
                <div class="box-header with-border">
                    <h3 class="box-title">Results</h3>
                </div>
                <!-- /.box-header -->
                <div class="box-body">
                    <div id="visualization-body"></div>
                </div>
                <!-- /.box-body -->
            </div>
            <!-- /.box -->
        </div>
        <!-- /.col -->
    </div>
    <!-- /.row -->
</section>
<!-- /.content -->

<script>
    function createVisualization(url){
        var container = new MultiVis("#visualization-body");
        $.ajax({
            url: url,
            beforeSend: function (request) {
                request.setRequestHeader("Authorization", 'Token'+' '+getAuthenticationCookie());
            },
            success: function(data) {
                //TODO Remove In Production
                console.log(data);

                // Add the newick data to generate a tree
                container.newickData = Newick.parse(data["newick"]);
                // Set the root of the tree
                container.newickRoot = container.newickData;

                // Add Genomes to Visualization
                for (var genomeIndex in container.newickData["branchset"]){
                    var genomeId = container.newickData["branchset"][genomeIndex]["name"];
                    var currentSeq = container.backbone.addSequence(
                        genomeId,
                        data["genomes"][genomeId]["length"],
                        data["genomes"][genomeId]["name"],
                        data["genomes"][genomeId]["name"]
                    );
                    // at this scale, individual scaling for sequences may not be usable...so used fixed scale
                    currentSeq.updateScale(0, container.getLargestSequenceSize(), container.getLargestSequenceSize());
                }

                // Add homolous regions to Visualization
                // TODO Fix ids for homologous region, doesnt use genome ids but rather goes from 0 -> N
                for (var i = 0; i < container.newickData["branchset"].length - 1; i++){
                    for (var j = 0; j < data["alignment"][i].length; j++){
                        container.backbone.addHomologousRegion(
                            i,
                            i + 1,
                            data["alignment"][i][j][0],
                            data["alignment"][i][j][1],
                            data["alignment"][i][j][2],
                            data["alignment"][i][j][3]
                        )
                    }
                }

                // render the graph
                container.render();
            }
        })
    }
</script>