**Be sure to bookmark this page to return to your work. The above URL is unique to you, and will be retained for three months following your last activity.**

:   Drag and drop your files into the 'Input datasets' box on the right. Alternatively, click the upload button <i class="icon-file-upload"></i> and select your datasets to upload.
    **Permitted dataset formats are Genbank (full) or EMBL and must include the genome sequence and annotated coding sequences.**

Select your data by clicking them
:   Hold Ctrl (⌘ for mac) to select multiple. Hold Shift to select a range. Ctrl-a (⌘-a for mac) to select all.

Upload a phylogenetic tree in newick format (Optional)
:   The tree will decide the order of alignment in the visualization. The file must have a '.newick' extension and refer to the genomes by accession. If no tree is provided, one will be automatically computed.

Select a reference genome (Optional)
:   If you are analysing draft genomes, select a reference genome that is closest to your datasets species. The drafts will be stitched in the order of alignment to the reference. If no reference is selected the draft contigs will be stitched in the order they appear in the uploaded dataset. The dataset label will be used for the stitched sequence identifier. If you specify a newick file, it must include this identifier.

Specify a label to identify the analysis, and click submit
:   The pending job will appear in the Recent Jobs tab. Once complete a "Visualize" button will appear in the Job History page along with the option to download the analysis.
